torch/backends/_nnapi/serializer.py

```diff
@@ -0,0 +1,1131 @@
+
+        inp_dim_orders = []
+        out_dim_orders = []
+
+        self_jitval = next(model.graph.inputs())
+        self.add_constant_value(self_jitval, self_jitval.type(), model)
+
+        for input_value, input_tensor in zip(list(model.graph.inputs())[1:], inputs):
+            op_id = self.add_tensor_operand_for_input(input_value, input_tensor)
+            inp_dim_orders.append(self.operands[op_id].dim_order.value)
+
+        for idx, node in enumerate(model.graph.nodes()):
+            LOG.debug("Processing node #%d: %r", idx, node)
+            self.add_node(node)
+
+        retn = model.graph.return_node()
+        assert retn.inputsSize() == 1
+        assert retn.outputsSize() == 0
+        # TODO: Make outputs a local variable? 
+        # TODO: Handle tuple-of-tensor return
+        for idx in range(1):
+            op_id = self.jitval_operand_map[retn.inputsAt(0)]
+            self.outputs.append(op_id)
+            out_dim_orders.append(self.operands[op_id].dim_order.value)
+
+        model = []
+
+        version = 1
+        header = struct.pack(
+                "iiiiii",
+                version,
+                len(self.operands),
+                len(self.values),
+                len(self.operations),
+                len(self.inputs),
+                len(self.outputs),
+                )
+        model.append(header)
+
+        serialized_values, serialized_value_data = self.serialize_values()
+
+        model.extend(struct.pack("iifi", t, len(d), s, z) for (t,d,_m,s,z) in self.operands)
+        model.extend(serialized_values)
+        model.extend(struct.pack("iii", *x) for x in self.operations)
+        model.extend(self.serialize_ints(fix_shape(dims, mf)) for (_, dims, mf, _, _) in self.operands)
+        model.extend(serialized_value_data)
+        model.append(self.serialize_ints(self.operation_args))
+        model.append(self.serialize_ints(self.inputs))
+        model.append(self.serialize_ints(self.outputs))
+
+        #return (b"".join(model), self.used_weight_tensor_names)
+        return (b"".join(model), self.used_weights, inp_dim_orders, out_dim_orders)
+
+
+    def serialize_values(self):
+        serialized_values = []
+        serialized_value_data = []
+        assert len(self.values) == len(self.value_data)
+        for ((op_index, source_type), data) in zip(self.values, self.value_data):
+            source_length = len(data)
+
+            # Pad with 0 bytes out to a multiple of 4 for alignment.
+            physical_length = ((source_length - 1) | 0x3) + 1
+            padded_data = data + (b"\0" * (physical_length - source_length))
+
+            serialized_values.append(struct.pack("iii", op_index, source_type, source_length))
+            serialized_value_data.append(padded_data)
+
+        return serialized_values, serialized_value_data
+
+
+    @staticmethod
+    def serialize_ints(ints):
+        return struct.pack("i" * len(ints), *ints)
+
+
+    ADDER_MAP = {
```

Comment:
Okay so these are the aten operators supported for the NNAPI execution. I'm guessing this list is dictated by the operation codes for the operators from https://developer.android.com/ndk/reference/group/neural-networks#operationcode

Reply:
Yeah, this maps from the internal TorchScript operator name to the function that needs to handle that node.  In some cases (like processing constants), there's no NNAPI operation generated.

Reply:
I'm assuming this map could grow if we are supporting more and more nnapi operations?

Reply:
Yes.  Though I'd like to replace the explicit map with a decorator.